diff --git a/src/ExternalEntityAccessControlHandler.php b/src/ExternalEntityAccessControlHandler.php
index 7c0d8b6..255e892 100755
--- a/src/ExternalEntityAccessControlHandler.php
+++ b/src/ExternalEntityAccessControlHandler.php
@@ -47,6 +47,9 @@ class ExternalEntityAccessControlHandler extends EntityAccessControlHandler {
         ->addCacheableDependency($this->entityType)
         ->addCacheableDependency($external_entity_type);
     }
+    elseif ($external_entity_type) {
+      $result = AccessResult::allowedIfHasPermission($account, "create {$external_entity_type->id()} external entity");
+    }
 
     return $result;
   }
